x-pack/plugins/osquery/server/plugin.ts

```diff
@@ -21,6 +22,158 @@ import { initSavedObjects } from './saved_objects';
 import { initUsageCollectors } from './usage';
 import { OsqueryAppContext, OsqueryAppContextService } from './lib/osquery_app_context_services';
 import { ConfigType } from './config';
+import { packSavedObjectType, savedQuerySavedObjectType } from '../common/types';
+
+const registerFeatures = (features: SetupPlugins['features']) => {
+  features.registerKibanaFeature({
+    id: 'osquery',
+    name: i18n.translate('xpack.features.osqueryFeatureName', {
+      defaultMessage: 'Osquery',
+    }),
+    order: 1300,
+    category: DEFAULT_APP_CATEGORIES.management,
+    app: ['osquery', 'kibana'],
+    catalogue: ['osquery'],
+    privileges: {
+      all: {
+        app: ['osquery', 'kibana'],
+        catalogue: ['osquery'],
+        savedObject: {
+          all: ['search', 'query'],
+          read: ['index-pattern'],
+        },
+        ui: ['show', 'save'],
+      },
+      read: {
+        app: ['osquery', 'kibana'],
+        catalogue: ['osquery'],
+        savedObject: {
+          all: [],
+          read: ['index-pattern', 'search', 'query'],
+        },
+        ui: ['show'],
+      },
+    },
+    subFeatures: [
+      {
+        name: i18n.translate('xpack.features.ossFeatures.discoverShortUrlSubFeatureName', {
+          defaultMessage: 'Live queries',
+        }),
+        privilegeGroups: [
+          {
+            groupType: 'mutually_exclusive',
+            privileges: [
+              {
+                id: 'live_queries_all',
+                includeIn: 'all',
+                name: 'All',
+                savedObject: {
+                  all: [],
+                  read: [],
+                },
+                ui: ['crudLiveQueries', 'readLiveQueries'],
+              },
+              {
+                id: 'live_queries_read',
+                includeIn: 'read',
+                name: 'Read',
+                savedObject: {
+                  all: [],
+                  read: [],
+                },
+                ui: ['readLiveQueries'],
+              },
+            ],
+          },
+          {
+            groupType: 'independent',
+            privileges: [
+              {
+                id: 'saved_queries_only',
+                name: i18n.translate(
+                  'xpack.features.ossFeatures.discoverCreateShortUrlPrivilegeName',
+                  {
+                    defaultMessage: 'Run Saved queries only',
+                  }
+                ),
+                includeIn: 'none',
+                savedObject: {
+                  all: [],
+                  read: [savedQuerySavedObjectType],
+                },
+                ui: ['savedQueriesOnly'],
+              },
+            ],
+          },
```

Comment:
Kibana's philosophy around privileges is that they are _additive_. What we have here appears to be intended to be the inverse, a _subtractive_ sub-feature privilege.

Perhaps you can turn this on its head and make it a sub-feature privilege that allows the user to "Run dynamic queries", which is included in the "read" feature privilege by default (`includeIn: 'read'`). Then administrators can optionally create a custom role and _exclude_ this sub-feature privilege from a user. WDYT?

If my assumption is correct, then you should also remove `savedQuerySavedObjectType` here, since you grant access to that saved object type in another sub-feature privilege below.